Full draft of FFNetAgent and supporting classes.

asmaalrawi opened the pull request:
- Relates to #2617 
- Addresses #2693 
- Replaces #2618 
- Replaces #2628
- Replaces #2506 
- Replaces #2759 
- Replaces #2629  

-  Adds classes: TensorOneD, TensorTwoD, TensorMath, LocalNeuralNet, DomainNetMap, RegionHintRecommender
- Adds agent FFNetAgent for steering CPU / GPU frequency decisions

alawibaba committed:
Full draft of FFNetAgent and supporting classes.

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,82 @@
+    class PlatformIO;
+
+    /// @brief Agent
+    class FFNetAgent : public geopm::Agent
```
Specifying the `geopm::` namespace is not needed here, since this class is within the `geopm` namespace.

Same note applies in other places as well.

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,82 @@
+                ///        performance for energy efficiency. A value of
+                ///        0 indicates an extreme preference for
+                ///        performance and a value of 10 indicates an
+                ///        extreme preference for energy efficiency.
```
Since there are other agents that accept a phi value in their policy, we should standardize it to provide a consistent interface. The other agents that accept phi accept a value between 0 and 1. Would this be feasible for this agent?

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+        }
+
+        for (int idx=0; idx<m_platform_topo.num_domain(GEOPM_DOMAIN_PACKAGE); ++idx) {
+            m_freq_control[GEOPM_DOMAIN_PACKAGE].push_back(m_platform_io.push_control("CPU_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_PACKAGE, idx));
```
nit: look into how to split this across multiple lines properly (it seems like it goes over our column limit)

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+        }
+
+        for (int idx=0; idx<m_platform_topo.num_domain(GEOPM_DOMAIN_GPU); ++idx) {
+            m_freq_control[GEOPM_DOMAIN_GPU].push_back(m_platform_io.push_control("GPU_CORE_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_GPU, idx));
```
Same nit as before regarding column limit

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    void FFNetAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+            std::vector<double>& out_sample)
```
nit: alignment

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+
+    // Distribute incoming policy to children
+    void FFNetAgent::split_policy(const std::vector<double>& in_policy,
+            std::vector<std::vector<double> >& out_policy)
```
nit: alignment

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+            }
+            if (m_phi_idx > 10) {
+                m_phi_idx = 10;
+            }
```
I would recommend checking this and throwing an exception in `validate_policy()` to ensure the user is aware of how to use phi and avoid hiding a performance issue.

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+        for (int domain_type : m_domain_types) {
+            m_net_map[domain_type] = geopm::make_unique<DomainNetMap>(
+                            m_platform_io,
+                            m_platform_topo.num_domain(domain_type));
```
nit: alignment

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+    // Read signals from the platform and calculate samples to be sent up
+    void FFNetAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        m_sample ++;
```
nit:
```suggestion
        ++m_sample;
```

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+        do {
+            geopm_time(&current_time);
+        }
+        while(geopm_time_diff(&m_last_wait, &current_time) < M_WAIT_SEC);
```
nit:
```suggestion
        while (geopm_time_diff(&m_last_wait, &current_time) < M_WAIT_SEC);
```

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,264 @@
+    {
+        std::vector<std::string> tracelist;
+        for (int domain_type : m_domain_types) {
+            std::vector<std::string> domain_tracenames = m_net_map.at(domain_type)->trace_names(std::to_string(domain_type));
```
nit: does this line go over our column limit?

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,82 @@
+            int m_sample;
+            bool m_has_gpus;
+            std::map<int, std::unique_ptr<DomainNetMap> > m_net_map;
+            std::map<int, std::unique_ptr<RegionHintRecommender> > m_freq_recommender;
```
Q: how are you planning on "injecting" the `DomainNetMap` and `RegionHintRecommender` objects to fully isolate `FFNetAgent` to unit test it?

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,231 @@
+#include "TensorOneD.hpp"
+#include "TensorTwoD.hpp"
+
+#define MAX_NNET_SIZE 1024*1024
```
Recommendation: make this a member of `DomainNetMap` and define it as a `constexpr` rather than a macro.

avilcheslopez commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,97 @@
+
+            geopm::PlatformIO &m_platform_io;
+
+            struct signal
```
Naming convention: use a "_s" suffix for struct names.

Same note applies to other structs.

avilcheslopez commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,97 @@
+            /// @return A map of string, float containing region class and probabilities
+            std::map<std::string, float> last_output(int index) const;
+        private:
+            std::shared_ptr<DenseLayer> json_to_DenseLayer(json11::Json obj) const;
```
Internally, this class depends on DenseLayer. Since these objects will be constructed internally after parsing the neural net configuration file, how are you planning on injecting this dependency to better isolate the class for unit testing?

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,232 @@
+        std::vector<std::shared_ptr<DenseLayer> > layers;
+        for (std::size_t layer_idx = 0; layer_idx < nnet_json["layers"].array_items().size(); ++layer_idx) {
+            layers.push_back(json_to_DenseLayer(nnet_json["layers"][layer_idx]));
+        }
```
I know the `json_to_*` functions perform part of the JSON input validation, but before calling them, there are some structural checks that should be made as well. For example, ensuring `nnet_json` contains "layers" and that it's an array.

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,82 @@
+
+        json11::Json fmap_json = json11::Json::parse(fbuf, err);
+
+        std::cout << "reading " << fmap_json.object_items().size() << " rows " << std::endl;
```
Debug comment?

cmcantalupo commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,102 @@
+            DomainNetMap(geopm::PlatformIO &plat_io, int count);
+            /// @brief Constructor with size specified 
+            ///
+            /// @param [in] n Size of 1D tensor
```
move above n -> count

cmcantalupo commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,102 @@
+            ///
+            /// @param [in] nn_path Path to neural net
+            /// @param [in] domain_type Domain type, defined by geopm_domain_e enum 
+            void load_neural_net(char* nn_path, int domain_type);
```
use const std::string & not char *

cmcantalupo commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,102 @@
+            void load_neural_net(char* nn_path, int domain_type);
+            /// @brief Calls sample_one for every domain index
+            void sample();
+            /// @brief Collects latest signals for a specific domain index and applies the 
```
Collects -> Sample

cmcantalupo commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,102 @@
+            ///        resulting TensorOneD state to the neural net.
+            ///
+            /// @param [in] domain_index
+            void sample_one(int index);
```
sample_one -> void compute_output(void)

cmcantalupo commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,102 @@
+            /// @brief generates the names for trace columns from the appropriate field in the neural net
+            ///
+            /// @param [in] domain String containing domain name. Must match neural net index
+            std::vector<std::string> trace_names(std::string domain) const;
```
Leave prefixing functionality to the agent

cmcantalupo commented on src/RegionHintRecommender.hpp:
```diff
@@ -0,0 +1,46 @@
+            ///        of region class string to double
+            ///
+            /// @param [in] fmap_path string containing path to freq map json
+            void load_freqmap(char* fmap_path);
```
char * -> const std::string &

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,232 @@
+        m_delta_inputs.resize(m_count);
+    }
+
+    std::shared_ptr<DenseLayer> DomainNetMap::json_to_DenseLayer(json11::Json obj) const {
```
For all JSON input, if you have schemas, please put them in: `<base>/json_schemas`. Also, include that schema in the agent's documentation. See `<base>/service/docs/source/geopmagent.1.rst` for an example of how it's included.

cmcantalupo replied:
-> const json11::Json &obj

cmcantalupo commented on src/RegionHintRecommender.hpp:
```diff
@@ -0,0 +1,46 @@
+            void set_max_freq(int max_freq);
+            /// @brief Set minimum frequency to recommend
+            /// @param [in] min_freq integer containing minimum frequency
+            void set_min_freq(int min_freq);
```
Add as constructor arguments

cmcantalupo commented on src/RegionHintRecommender.hpp:
```diff
@@ -0,0 +1,46 @@
+    class RegionHintRecommender
+    {
+        public:
+            /// @brief Loads frequency map file from json11 into a std::map
```
Add make_unique()
RegionHintRecommender(const std::string &fmap_path, min_freq, max_freq)
virtual ~RegionHintRecommender()

cmcantalupo commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,232 @@
+        m_last_output[domain_index] = m_neural_net->forward(xs);
+    }
+
+    std::vector<std::string> DomainNetMap::trace_names(std::string domain) const
```
const std::string &domain

cmcantalupo commented on src/DenseLayerImp.hpp:
```diff
@@ -0,0 +1,51 @@
+            /// @throws geopm::Exception if input vector is incompatible with the DenseLayer
+            ///
+            /// @returns Returns a TensorOneD object of output values
+            TensorOneD forward(const TensorOneD input) const override;
```
const TensorOneD &input

cmcantalupo commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,232 @@
+        : m_platform_io(plat_io)
+          , m_count(count)
+    {
+        m_last_output.resize(m_count);
+        m_signal_inputs.resize(m_count);
+        m_delta_inputs.resize(m_count);
```
Add these to initializer list

Commit:
Changed DomainNetMap to have one instance per domain, rather than one instance per domain type, and modified FFNetAgent accordingly.

Commit:
Changed RegionHintRecommender's interface to use a double for phi (0 - 1) rather than a map-specific index.

Commit:
Modified RegionHintRecommender to follow interface/implementation class pattern.

alawibaba committed:
Added unit tests and minor cleanup for TensorTwoD.

Commit:
- Added destructor to RegionHintRecommender
- Reordered private variables in RegionHintRecommenderImp
- Made iterator variable a reference to avoid unnecessary copies.
- Copying math instance for both Tensor classes with equal operator override.
- Using initializer list for DenseLayer constructor.
- Fix some TensorMath object leaks in TensorOneD

Commit:
- Added Mocks for DomainNetMap and RegionHintRecommender.
- Added test constructor to FFNetAgent
- Added a construct_and_init method to FFNetAgentTest

Co-Authored-By: Ali Mohammad <ali.mohammad@intel.com>
Co-Authored-By: Asma H Al-Rawi <asma.h.al-rawi@intel.com>

alawibaba committed:
Added virtual destructor to TensorOneD and TensorTwoD which fixed the leaking mock objects in the tests.

Commit:
Removed PlatformIO from DomainNetMap constructor arg.

Signed-off-by: Ali Mohammad<ali.mohammad@intel.com>

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,112 @@
+            std::vector<geopm_domain_e> m_domain_types;
+            std::vector<m_domain_key_s> m_domains;
+
+            static const std::map<geopm_domain_e, const char *> nnet_envname;
```
Nit: naming convention:
- Upper case for constants
- M_ prefix for class members

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,363 @@
+                        domain_key.type,
+                        domain_key.index)
+                    ));
+        }
```
Q: why not leverage `std::map`'s copy constructor? I see the use of `std::unique_ptr` as the element type might cause some issues. Would it make sense to use `std::shared_ptr`, instead?

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,363 @@
+            const geopm::PlatformTopo &topo,
+            std::map<std::pair<geopm_domain_e, int>, std::unique_ptr<DomainNetMap> > &net_map,
+            std::map<geopm_domain_e, std::unique_ptr<RegionHintRecommender> > &freq_recommender
+            )
```
Make this the "base" constructor called by the others. If the others don't need to pass a `DomainNetMap` nor a `RegionHintRecommender`, they can simply pass an empty map (`{ }`). If the maps are empty, then the constructor can proceed with the logic to load the neural nets from configuration.

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,363 @@
+    // Validate incoming policy and configure default policy requests.
+    void FFNetAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        if(in_policy.size() != M_NUM_POLICY) {
```
Nit
```suggestion
        if (in_policy.size() != M_NUM_POLICY) {
```

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,363 @@
+        return false;
+    }
+
+    void FFNetAgent::adjust_platform(const std::vector<double>& in_policy)
```
Nit:
```suggestion
    void FFNetAgent::adjust_platform(const std::vector<double> &in_policy)
```

Commit:
Added DomainNetMap neural net schema.

Commit:
Replaced direct construction with factory in DomainNetMap, with corresponding additional classes.

alawibaba committed:
- Created a gtest matcher for Tensor instances that does not rely on gtest's comparisons which cause memory leaks with shared_ptr of mocks.
- Fixed move constructor for TensorOneD.

Commit:
Removed std::move from temp object in DomainNetMap

alawibaba committed:
Added tests for DenseLayer.

Commit:
Changed make_uniq to make_shared in FFNetAgent to allow tests to access
DomainNetMap and RegionHintRecommender objects.

Commit:
Added adjust platform test to FFNetAgentTest

alawibaba committed:
- Added mocks for DenseLayer, LocalNeuralNet and NNFactory.
- Completed FFNetAgentTest
- Removed unused variables and fixed a type in FFNetAgentTest.

Commit:
Added chapter7 man page for ffnet and adjusted policy name.

alawibaba committed:
- Added DomainNetMap unit tests, made corresponding modifications to interfaces and mocks to enable testing.
- Added unit tests for RegionHintRecommender.

alawibaba committed:
- Added new json schema to license exempt list.
- Fixing typos and license stuff.
- Adding DomainNetMap schema to manifest.
- Added new manpage to manifest.

alawibaba committed:
Added LocalNeuralNetTest.

alawibaba committed:
Fixed persistent memory leak issue with DenseLayerTest!

alawibaba committed:
Added json schema for RegionHintRecommender's frequency map.

alawibaba committed:
Switched neural net library from float to double and propagated the change to tests.

Commit:
Updated some doxygen strings to correspond to recent code changes.

Commit:
- Fixed DenseLayerTest to use SetUp without running into memory leaks.
- Moved repeated strut code into SetUp for several tests.

Commit:
Improving tests and schema for neural net:

- Improved DomainNetMap neural net json schema
- Added json checks to DomainNetMap and RegionHintRecommender
- Added tests for json format to DomainNetMap and RegionHintRecommender
- Added doxygen strings to LocalNeuralNet

Commit:
- Cleaned up unnecessary geopm:: and various minor formatting changes.
- Replaced "probabilities" with "logits" in doxygen for clarification.
- Minor code clean-up for NN code

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+
+        for (geopm_domain_e domain_type : m_domain_types) {
+            m_freq_recommender[domain_type] = std::make_shared<RegionHintRecommenderImp>(
+                    getenv(freqmap_envname.at(domain_type)),
```
Calling getenv() is generally to be avoided if possible.  Why is more than one environment variable needed if there is a requirement?  Is this something that could be managed by the Environment class?

cmcantalupo commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        return geopm::make_unique<RegionHintRecommenderImp>(fmap_path, min_freq, max_freq);
+    }
+
+    RegionHintRecommenderImp::RegionHintRecommenderImp(const std::string fmap_path, int min_freq, int max_freq)
```
```suggestion
    RegionHintRecommenderImp::RegionHintRecommenderImp(const std::string &fmap_path, int min_freq, int max_freq)
```

cmcantalupo commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+
+namespace geopm
+{
+    std::unique_ptr<RegionHintRecommender> RegionHintRecommender::make_unique(const std::string fmap_path, int min_freq, int max_freq)
```
```suggestion
    std::unique_ptr<RegionHintRecommender> RegionHintRecommender::make_unique(const std::string &fmap_path, int min_freq, int max_freq)
```

cmcantalupo commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+
+    double RegionHintRecommenderImp::recommend_frequency(std::map<std::string, double> nn_output, double phi) const {
```
```suggestion
    double RegionHintRecommenderImp::recommend_frequency(const std::map<std::string, double> &nn_output, double phi) const {
```

cmcantalupo commented on src/RegionHintRecommender.hpp:
```diff
@@ -0,0 +1,41 @@
+            /// @param [in] fmap_path string containing path to freq map json
+            /// @param [in] min_freq integer containing minimum frequency
+            /// @param [in] max_freq integer containing maximum frequency
+            static std::unique_ptr<RegionHintRecommender> make_unique(const std::string fmap_path, int min_freq, int max_freq);
```
```suggestion
            static std::unique_ptr<RegionHintRecommender> make_unique(const std::string &fmap_path, int min_freq, int max_freq);
```

cmcantalupo commented on src/RegionHintRecommender.hpp:
```diff
@@ -0,0 +1,41 @@
+            /// @param [in] nn_output Output from the DomainNetMap neural net evaluation, as a map from
+            ///             trace names to doubles.
+            /// @param [in] phi User-input perf-energy bias
+            virtual double recommend_frequency(std::map<std::string, double> nn_output, double phi) const = 0;
```
```suggestion
            virtual double recommend_frequency(const std::map<std::string, double> &nn_output, double phi) const = 0;
```

cmcantalupo commented on src/RegionHintRecommenderImp.hpp:
```diff
@@ -0,0 +1,35 @@
+            /// @param [in] nn_output List of region classification names and
+            ///             logits output from region class neural net
+            /// @param [in] phi User-input perf-energy bias
+            double recommend_frequency(std::map<std::string, double> nn_output, double phi) const override;
```
```suggestion
            double recommend_frequency(const std::map<std::string, double> &nn_output, double phi) const override;
```

cmcantalupo commented on src/TensorTwoD.cpp:
```diff
@@ -0,0 +1,153 @@
+        set_data(input);
+    }
+
+    TensorTwoD::TensorTwoD(std::vector<std::vector<double>> input)
```
```suggestion
    TensorTwoD::TensorTwoD(const std::vector<std::vector<double>> &input)
```

cmcantalupo commented on src/TensorTwoD.cpp:
```diff
@@ -0,0 +1,153 @@
+    {
+    }
+
+    TensorTwoD::TensorTwoD(std::vector<std::vector<double>> input,
```
```suggestion
    TensorTwoD::TensorTwoD(const std::vector<std::vector<double> > &input,
```

cmcantalupo commented on src/TensorTwoD.cpp:
```diff
@@ -0,0 +1,153 @@
+        return m_mat;
+    }
+
+    void TensorTwoD::set_data(const std::vector<TensorOneD> data)
```
```suggestion
    void TensorTwoD::set_data(const std::vector<TensorOneD> &data)
```

cmcantalupo reviewed (commented):
I started to review for functions that copy complex object on calling.  This pattern is throughout this patch.  Please address all of the functions that are marked, but also the others where complex objects are not passed by reference.

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+#include "RegionHintRecommender.hpp"
+#include "DomainNetMapImp.hpp"
+#include "RegionHintRecommenderImp.hpp"
+#include "geopm/PlatformTopo.hpp"
```
Move this to DomainNetMap.cpp

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+#include "Agent.hpp"
+#include "geopm_time.h"
+#include "DomainNetMap.hpp"
+#include "RegionHintRecommender.hpp"
```
Remove

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+#include "geopm_time.h"
+#include "DomainNetMap.hpp"
+#include "RegionHintRecommender.hpp"
+#include "DomainNetMapImp.hpp"
```
Remove

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+            void init(int level, const std::vector<int> &fan_in, bool is_level_root) override;
+            void validate_policy(std::vector<double> &in_policy) const override;
+            void split_policy(const std::vector<double> &in_policy,
+                    std::vector<std::vector<double> > &out_policy) override;
```
nit: alignment

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+                    std::vector<std::vector<double> > &out_policy) override;
+            bool do_send_policy(void) const override;
+            void aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                    std::vector<double> &out_sample) override;
```
nit: alignment

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+            static const std::map<geopm_domain_e, std::string> min_freq_control_name;
+            static const std::map<geopm_domain_e, std::string> trace_suffix;
+
+            void init_domain_indices(const PlatformTopo &topo);
```
nit: move up to where the other method is in the private section

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+
+using geopm::Agent;
+using geopm::PlatformIO;
+using geopm::PlatformTopo;
```
Remove (this class is in the geopm namespace)

avilcheslopez commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,114 @@
+
+            PlatformIO &m_platform_io;
+            geopm_time_s m_last_wait;
+            const double M_WAIT_SEC;
```
Could be `static constexpr`

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+            PlatformIO &plat_io,
+            const PlatformTopo &topo,
+            std::map<std::pair<geopm_domain_e, int>, std::shared_ptr<DomainNetMap> > &net_map,
+            std::map<geopm_domain_e, std::shared_ptr<RegionHintRecommender> > &freq_recommender
```
const map&

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+        geopm_time(&m_last_wait);
+        init_domain_indices(topo);
+
+        for (const geopm_domain_e domain_type : m_domain_types) {
+            m_freq_recommender[domain_type] = freq_recommender.at(
+                        domain_type
+                    );
+        }
+
+        for (const m_domain_key_s domain_key : m_domains) {
+            m_net_map[domain_key] = net_map.at(
+                    std::make_pair(
+                        domain_key.type,
+                        domain_key.index)
+                    );
+        }
+    }
```
Q: is there a reason why you didn't simply use `std::map::operator=`?

avilcheslopez replied:
Asma: yes, use `operator=

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+
+        //Loading neural nets
+        for (const m_domain_key_s domain_key : m_domains) {
+            m_net_map[domain_key] = std::make_shared<DomainNetMapImp>(
```
Instead of directly instantiating a `DomainNetMapImp` object, this should use a factory pattern/method for requesting an object of type `DomainNetMap`.

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+        }
+
+        for (geopm_domain_e domain_type : m_domain_types) {
+            m_freq_recommender[domain_type] = std::make_shared<RegionHintRecommenderImp>(
```
Instead of directly instantiating a `RegionHintRecommenderImp` object, this should use a factory pattern/method for requesting an object of type `RegionHintRecommender`.

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+
+        for (const m_domain_key_s domain_key : m_domains) {
+            m_net_map[domain_key]->sample();
+        }
```
```suggestion
        for (auto &nm : m_net_map) {
            nm.second()->sample();
        }
```

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+            for (std::size_t idx=0; idx < domain_row.size(); ++vidx, ++idx) {
+                values[vidx] = domain_row[idx];
+            }
+        }
```
```suggestion
        for (const auto &nm : m_net_map) {
            std::vector<double> domain_row = nm.second()->trace_values();
            for (std::size_t idx=0; idx < domain_row.size(); ++vidx, ++idx) {
                values[vidx] = domain_row[idx];
            }
        }
```

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,362 @@
+
+    // Distribute incoming policy to children
+    void FFNetAgent::split_policy(const std::vector<double>& in_policy,
+                                  std::vector<std::vector<double> >& out_policy)
```
nit:
```suggestion
    void FFNetAgent::split_policy(const std::vector<double> &in_policy,
                                  std::vector<std::vector<double> > &out_policy)
```

avilcheslopez commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,60 @@
+
+#include <memory>
+
+#include "geopm/PlatformTopo.hpp"
```
Move to DomainNetMap.cpp

avilcheslopez commented on src/DomainNetMapImp.hpp:
```diff
@@ -0,0 +1,85 @@
+        public:
+            DomainNetMapImp(const std::string nn_path, geopm_domain_e domain_type, int domain_index);
+
+            DomainNetMapImp(const std::string nn_path, geopm_domain_e domain_type, int domain_index, PlatformIO &plat_io, std::shared_ptr<NNFactory> nn_factory);
```
Check line length

avilcheslopez replied:
```suggestion
            DomainNetMapImp(const std::string &nn_path, geopm_domain_e domain_type, int domain_index, PlatformIO &plat_io, std::shared_ptr<NNFactory> nn_factory);
```

avilcheslopez commented on src/DomainNetMapImp.hpp:
```diff
@@ -0,0 +1,85 @@
+                double value;
+            };
+
+            static constexpr int m_max_nnet_size = 1024;
```
Naming convention: uppercase

avilcheslopez replied:
Naming convention: uppercase

avilcheslopez replied:
Naming convention: uppercase

avilcheslopez commented on src/DomainNetMapImp.hpp:
```diff
@@ -0,0 +1,85 @@
+#include <memory>
+
+#include "geopm/json11.hpp"
+#include "geopm/PlatformIO.hpp"
```
Move to DomainNetMap.cpp

avilcheslopez commented on src/DomainNetMap.hpp:
```diff
@@ -0,0 +1,60 @@
+            ///
+            /// @throws geopm::Exception if neural net file does not contain either a non-empty 
+            ///         array "signal_inputs" or a non-empty array "delta_inputs"
+            static std::unique_ptr<DomainNetMap> make_unique(const std::string n_path, geopm_domain_e domain_type, int domain_index);
```
```suggestion
            static std::unique_ptr<DomainNetMap> make_unique(const std::string &n_path, geopm_domain_e domain_type, int domain_index);
```

avilcheslopez commented on src/DomainNetMapImp.hpp:
```diff
@@ -0,0 +1,85 @@
+    class DomainNetMapImp : public DomainNetMap
+    {
+        public:
+            DomainNetMapImp(const std::string nn_path, geopm_domain_e domain_type, int domain_index);
```
```suggestion
            DomainNetMapImp(const std::string &nn_path, geopm_domain_e domain_type, int domain_index);
```

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+            const std::string nn_path,
+            geopm_domain_e domain_type,
+            int domain_index)
+        : DomainNetMapImp(nn_path, domain_type, domain_index, platform_io(), std::make_shared<NNFactoryImp>()) {
```
Use the factory method (`NNFactory::make_unique()`) instead of directly instantiating the NNFactory type.

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        std::streampos length = file.tellg();
+        file.seekg(0, std::ios::beg);
+
+        if (length >= m_max_nnet_size) {
+            throw Exception("Neural net file exceeds maximum size.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
```
Q: what if file is empty?

avilcheslopez replied:
Asma: throw an error (empty file not acceptable)

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+
+        if (!file) {
+            throw Exception("Unable to open neural net file.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        const json11::Json nnet_json = json11::Json::parse(buf, err);
+
+        // make sure that there no unexpected keys in the json
+        std::vector<std::string> expected_keys = {"layers", "signal_inputs", "delta_inputs", "trace_outputs", "description"};
```
Make this const, perhaps a static const property of the class.

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+
+        // make sure that there no unexpected keys in the json
+        std::vector<std::string> expected_keys = {"layers", "signal_inputs", "delta_inputs", "trace_outputs", "description"};
+        for (auto &key : nnet_json.object_items()) {
```
First check `nnet_json.is_object()`

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+            }
+        }
+
+        if (! nnet_json["layers"].is_array()) {
```
nit:
```suggestion
        if (!nnet_json["layers"].is_array()) {
```

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        }
+
+        if (nnet_json["signal_inputs"].array_items().size() == 0 
+            && nnet_json["delta_inputs"].array_items().size() == 0) {
```
First check if these objects are arrays

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        }
+        
+        std::vector<std::shared_ptr<DenseLayer>> layers;
+        for (std::size_t layer_idx = 0; layer_idx < nnet_json["layers"].array_items().size(); ++layer_idx) {
```
```suggestion
        for (const auto &layer : nnet_json["layers"]) {
```

Edit line below accordingly.

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+            layers.push_back(json_to_DenseLayer(nnet_json["layers"][layer_idx]));
+        }
+
+	if (layers.size() == 0) {
+            throw Exception(
+                    "Layers must be non-empty.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
```
Q: why not check this earlier when nnet_json["layers"] is checked?

avilcheslopez replied:
Asma: yeah, move it before

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	if (nnet_json["signal_inputs"].array_items().size()
+                + nnet_json["delta_inputs"].array_items().size()
+                != m_neural_net->get_input_dim()) {
```
nit: alignment

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+	}
+
+        if (nnet_json["trace_outputs"].array_items().size()
+                != m_neural_net->get_output_dim()) {
```
nit: alignment

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        for (std::size_t i=0; i<nnet_json["signal_inputs"].array_items().size(); i++) {
```
```suggestion
        for (const auto &signal : nnet_json["signal_inputs"]) {
```

Modify lines below accordingly.

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        for (std::size_t i=0; i<nnet_json["signal_inputs"].array_items().size(); i++) {
+            m_signal_inputs.push_back({
+                    m_platform_io.push_signal(
+                            nnet_json["signal_inputs"][i].string_value(),
```
Check array element type

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+                    NAN});
+        }
+
+        for (std::size_t i=0; i<nnet_json["delta_inputs"].array_items().size(); i++) {
```
```suggestion
        for (const auto &input : nnet_json["delta_inputs"]) {
```

Modify lines below accordingly

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        for (std::size_t i=0; i<nnet_json["delta_inputs"].array_items().size(); i++) {
+            m_delta_inputs.push_back({
+                    m_platform_io.push_signal(
+                            nnet_json["delta_inputs"][i][0].string_value(),
```
Check array item type

avilcheslopez replied:
Check `delta_inputs` elements are arrays of size 2

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+                            domain_type,
+                            domain_index),
+                    m_platform_io.push_signal(
+                            nnet_json["delta_inputs"][i][1].string_value(),
```
Check array item type

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+                    NAN, NAN, NAN, NAN});
+        }
+
+        for (std::size_t i=0; i<nnet_json["trace_outputs"].array_items().size(); i++) {
```
```suggestion
        for (const auto &output : nnet_json["trace_outputs"]) {
```

Modify line below accordingly

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+        std::vector<double> xs;
+
+        // Sample latest signal values
+        for (std::size_t i=0; i<m_signal_inputs.size(); i++) {
```
```suggestion
        for (const auto &inputs : m_signal_inputs) {
```

Modify lines below accordingly

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,265 @@
+            m_signal_inputs[i].signal = m_platform_io.sample(m_signal_inputs[i].batch_idx);
+            xs.push_back(m_signal_inputs[i].signal);
+        }
+        for (std::size_t i=0; i<m_delta_inputs.size(); i++) {
```
```suggestion
        for (const &auto inputs : m_delta_inputs) {
```

Modify lines below accordingly

avilcheslopez commented on src/DenseLayer.cpp:
```diff
@@ -0,0 +1,61 @@
+    {
+        if (input.get_dim() != m_weights.get_cols()) {
+            throw Exception("Input vector dimension is incompatible with network.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/DenseLayer.cpp:
```diff
@@ -0,0 +1,61 @@
+
+        if (weights.get_rows() != biases.get_dim()) {
+            throw Exception("Incompatible dimensions for weights and biases.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/DenseLayer.cpp:
```diff
@@ -0,0 +1,61 @@
+    {
+        if (weights.get_rows() == 0 && weights.get_cols() == 0) {
+            throw Exception("Empty array is invalid for neural network weights.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,66 @@
+
+#include "config.h"
+
+#include "TensorMath.hpp"
```
Is this needed here?

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,66 @@
+        for (std::size_t idx=1; idx < layers.size(); ++idx) {
+            if (layers[idx]->get_input_dim() != layers[idx-1]->get_output_dim()) {
+                throw Exception("Incompatible dimensions for consecutive layers.",
+                                       GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,66 @@
+    {
+        if (inp.get_dim() != m_layers[0]->get_input_dim()) {
+            throw Exception("Input vector dimension is incompatible with network.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    size_t LocalNeuralNetImp::get_output_dim() const {
+        return m_layers[m_layers.size() - 1]->get_output_dim();
```
Q: can `m_layers.size() == 0`? Is having 0 layers a valid use case?

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        std::ifstream ffile(fmap_path);
+        if (!ffile) {
+            throw Exception("Unable to open frequency map file.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        
+        if (fmap_json.is_null() || !fmap_json.is_object()) {
+            throw Exception("Frequency map file format is incorrect.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        json11::Json fmap_json = json11::Json::parse(fbuf, err);
+        
+        if (fmap_json.is_null() || !fmap_json.is_object()) {
+            throw Exception("Frequency map file format is incorrect.\n",
```
Suggestion: include error message returned from `Json::parse()` to exception error message.

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        for (const auto &row : fmap_json.object_items()) {
+            if (! row.second.is_array() || row.second.array_items().size() == 0) {
```
nit:
```suggestion
            if (!row.second.is_array() || row.second.array_items().size() == 0) {
```

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        for (const auto &row : fmap_json.object_items()) {
+            if (! row.second.is_array() || row.second.array_items().size() == 0) {
+                throw Exception("Frequency map file format is incorrect.\n",
+                                       GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+                                       GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            m_freq_map[row.first].resize(row.second.array_items().size());
+            for (int idx=0; idx < (int)m_freq_map[row.first].size(); idx++) {
```
```suggestion
            for (size_t idx=0; idx < m_freq_map[row.first].size(); ++idx) {
```

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+        double ZZ = 0;
+        double freq = 0;
+        for (const auto & [region_name, probability] : nn_output) {
+            int phi_idx = (int)(phi * (m_freq_map.at(region_name).size() - 1));
```
```suggestion
            size_t phi_idx = std::trunc(phi * (m_freq_map.at(region_name).size() - 1)); // or std::floor(), or std::round()?
```

avilcheslopez commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,91 @@
+            }
+        }
+
+        double ZZ = 0;
```
```suggestion
        double zz = 0;
```

Modify code below accordingly

avilcheslopez commented on src/NNFactory.cpp:
```diff
@@ -0,0 +1,39 @@
+
+#include "DenseLayerImp.hpp"
+#include "LocalNeuralNetImp.hpp"
+#include "NNFactory.hpp"
```
Can be removed

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,66 @@
+#include "TensorOneD.hpp"
+#include "TensorTwoD.hpp"
+#include "DenseLayer.hpp"
+#include "LocalNeuralNet.hpp"
```
Can be removed

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+            /// @param [in] other The summand
+            ///
+            /// @return Returns a 1D tensor, the sum of two 1D tensors
+            virtual TensorOneD add(const TensorOneD& tensor_a, const TensorOneD& tensor_b) const = 0;
```
nit:
```suggestion
            virtual TensorOneD add(const TensorOneD &tensor_a, const TensorOneD &tensor_b) const = 0;
```

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+            /// @param [in] other The subtrahend
+            ///
+            /// @return A 1D tensor, the difference of two 1D tensors.
+            virtual TensorOneD subtract(const TensorOneD& tensor_a, const TensorOneD& tensor_b) const = 0;
```
nit:
```suggestion
            virtual TensorOneD subtract(const TensorOneD &tensor_a, const TensorOneD &tensor_b) const = 0;
```

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+            /// @param [in] other The multiplicand
+            ///
+            /// @return Returns a 1D tensor, the product of two 1D tensors
+            virtual double inner_product(const TensorOneD& tensor_a, const TensorOneD& tensor_b) const = 0;
```
nit:
```suggestion
            virtual double inner_product(const TensorOneD &tensor_a, const TensorOneD &tensor_b) const = 0;
```

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+            virtual double inner_product(const TensorOneD& tensor_a, const TensorOneD& tensor_b) const = 0;
+            /// @brief Compute logistic sigmoid function of 1D Tensor
+
+            virtual TensorOneD sigmoid(const TensorOneD& tensor) const = 0;
```
nit:
```suggestion
            virtual TensorOneD sigmoid(const TensorOneD &tensor) const = 0;
```

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+            /// 
+            /// @throws geopm::Exception if the sizes are incompatible, i.e. if 2D
+            ///         tensor number of columns is unequal to 1D tensor number of rows
+            virtual TensorOneD multiply(const TensorTwoD&, const TensorOneD&) const = 0;
```
nit:
```suggestion
            virtual TensorOneD multiply(const TensorTwoD &, const TensorOneD &) const = 0;
```

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+            TensorOneD subtract(const TensorOneD& tensor_a, const TensorOneD& tensor_b) const override;
+            double inner_product(const TensorOneD& tensor_a, const TensorOneD& tensor_b) const override;
+            TensorOneD sigmoid(const TensorOneD& tensor) const override;
+            TensorOneD multiply(const TensorTwoD&, const TensorOneD&) const override;
```
nit:
```suggestion
            TensorOneD add(const TensorOneD &tensor_a, const TensorOneD &tensor_b) const override;
            TensorOneD subtract(const TensorOneD &tensor_a, const TensorOneD &tensor_b) const override;
            double inner_product(const TensorOneD &tensor_a, const TensorOneD &tensor_b) const override;
            TensorOneD sigmoid(const TensorOneD &tensor) const override;
            TensorOneD multiply(const TensorTwoD &, const TensorOneD &) const override;
```

avilcheslopez commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+        if (tensor_a.get_dim() != tensor_b.get_dim()) {
+            throw Exception("Adding vectors of mismatched dimensions.",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+        if (tensor_a.get_dim() != tensor_b.get_dim()) {
+            throw Exception("Subtracting vectors of mismatched dimensions.",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+        if (tensor_a.get_dim() != tensor_b.get_dim()) {
+            throw Exception("Inner product of vectors of mismatched dimensions.",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,100 @@
+    TensorOneD TensorMathImp::sigmoid(const TensorOneD& tensor) const
+    {
+        TensorOneD rval(tensor.get_dim());
+        for(std::size_t idx = 0; idx < tensor.get_dim(); idx++) {
```
nit:
```suggestion
        for (std::size_t idx = 0; idx < tensor.get_dim(); ++idx) {
```

avilcheslopez commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+        if (tensor_a.get_cols() != tensor_b.get_dim()) {
+            throw Exception("Attempted to multiply matrix and vector with incompatible dimensions.",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
nit: alignment

avilcheslopez commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,100 @@
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        const auto &mat = tensor_a.get_data();
```
nit: naming convention

avilcheslopez commented on src/TensorOneD.hpp:
```diff
@@ -0,0 +1,104 @@
+            /// @brief Overload = operator with an in-place deep copy
+            ///
+            /// @param [in] other The assignee (tensor to be copied)
+            TensorOneD& operator=(const TensorOneD& other);
```
nit:
```suggestion
            TensorOneD& operator=(const TensorOneD &other);
```

avilcheslopez commented on src/TensorOneD.hpp:
```diff
@@ -0,0 +1,104 @@
+            /// @brief Constructs a deep copy of the argument
+            ///
+            /// @param [in] TensorOneD Tensor to copy
+            TensorOneD(const TensorOneD&);
```
nit:
```suggestion
            TensorOneD(const TensorOneD &);
```

avilcheslopez commented on src/TensorMath.hpp:
```diff
@@ -0,0 +1,75 @@
+    /// @brief Class to perform operations on 1D and 2D Tensors,
+    ///        aka vectors and matrices, suitable for use in
+    ///        feed-forward neural networks.
+    class TensorMath
```
It seems `TensorMath` is missing a factory method

avilcheslopez commented on src/TensorOneD.cpp:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    TensorOneD::TensorOneD(std::vector<double> input)
+        : TensorOneD(input, std::make_shared<TensorMathImp>())
```
Should call `TensorMath` factory method

avilcheslopez commented on src/TensorOneD.cpp:
```diff
@@ -0,0 +1,115 @@
+        return m_vec == other.m_vec;
+    }
+
+    double& TensorOneD::operator[] (size_t idx)
```
nit:
```suggestion
    double& TensorOneD::operator[](size_t idx)
```

avilcheslopez commented on src/TensorOneD.cpp:
```diff
@@ -0,0 +1,115 @@
+        return m_vec.at(idx);
+    }
+
+    double TensorOneD::operator[] (size_t idx) const
```
nit:
```suggestion
    double TensorOneD::operator[](size_t idx) const
```

avilcheslopez commented on src/TensorTwoD.cpp:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    TensorTwoD::TensorTwoD(size_t rows, size_t cols)
+        : TensorTwoD(rows, cols, std::make_shared<TensorMathImp>())
```
Should call `TensorMath` factory method

avilcheslopez commented on src/TensorTwoD.cpp:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    TensorTwoD::TensorTwoD(std::vector<TensorOneD> input)
+        : TensorTwoD(input, std::make_shared<TensorMathImp>())
```
Should call `TensorMath` factory method

avilcheslopez commented on src/TensorTwoD.cpp:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    TensorTwoD::TensorTwoD(std::vector<std::vector<double>> input)
+        : TensorTwoD(input, std::make_shared<TensorMathImp>())
```
Should call `TensorMath` factory method

avilcheslopez commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,275 @@
+            geopm_domain_e domain_type,
+            int domain_index)
+        : DomainNetMapImp(nn_path, domain_type, domain_index, platform_io(),
+                          std::make_shared<NNFactoryImp>()) {
```
Call factory method

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,359 @@
+
+        if (freq_recommender.empty()) {
+            for (geopm_domain_e domain_type : m_domain_types) {
+                m_freq_recommender[domain_type] = std::make_shared<RegionHintRecommenderImp>(
```
Call factory method

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,359 @@
+
+        if (net_map.empty()) {
+            for (const m_domain_key_s domain_key : m_domains) {
+                m_net_map[domain_key] = std::make_shared<DomainNetMapImp>(
```
Call factory method

Commit:
Moved FFNet agent to beta flag.

avilcheslopez committed:
Add unit tests for some NN classes

Add unit tests for DomainNetMap, RegionHintRecommender,
LocalNeuralNet.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
- Minor improvements in DomainNetMap schema
- Include schemas in FFNetAgent documentation
- Fixed FFNetAgent man page build
- Minor clean-up/fixes/improvements in NN code and unit tests

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on test/DomainNetMapTest.cpp:
```diff
@@ -0,0 +1,498 @@
+
+void DomainNetMapTest::SetUp()
+{
+    fake_nn_factory = std::make_shared<NiceMock<MockNNFactory>>();
```
Why was NiceMock needed here? Were there issues setting up expectations such that suppressing warnings for "uninteresting calls" was preferred?

alawibaba replied:
Fixed

avilcheslopez commented on test/TensorTwoDTest.cpp:
```diff
@@ -0,0 +1,133 @@
+
+TEST_F(TensorTwoDTest, test_degenerate_size) {
+    TensorTwoD x;
+    EXPECT_EQ(0u, x.get_cols());
```
What should `get_rows()` return?

avilcheslopez commented on test/TensorTwoDTest.cpp:
```diff
@@ -0,0 +1,133 @@
+TEST_F(TensorTwoDTest, test_bad_dimensions)
+{
+    std::vector<std::vector<double> > vals = {{1}, {2, 3}};
+    EXPECT_THROW(new TensorTwoD(vals), geopm::Exception);  // TODO - new?
```
GEOPM_EXPECT_THROW?

avilcheslopez commented on test/TensorTwoDTest.cpp:
```diff
@@ -0,0 +1,133 @@
+TEST_F(TensorTwoDTest, test_empty_weights)
+{
+    std::vector<std::vector<double> > vals = {};
+    EXPECT_THROW(new TensorTwoD(vals), geopm::Exception);  // TODO - new?
```
GEOPM_EXPECT_THROW?

avilcheslopez commented on test/TensorTwoDTest.cpp:
```diff
@@ -0,0 +1,133 @@
+using ::testing::Return;
+using ::testing::_;
+
+class TensorTwoDTest : public ::testing::Test
```
Potential missing tests:
* `operator==()`
* `set_data()`

avilcheslopez commented on test/TensorOneDTest.cpp:
```diff
@@ -0,0 +1,66 @@
+{
+};
+
+TEST_F(TensorOneDTest, test_sum)
```
Potential missing tests:
* `TensorOneD(size_t)`
* Copy constructor
* `operator==()`
* `operator-()`
* `operator*()`
* `sigmoid()`

avilcheslopez commented on test/FFNetAgentTest.cpp:
```diff
@@ -0,0 +1,519 @@
+        m_net_map[std::make_pair(GEOPM_DOMAIN_PACKAGE, idx)]
+            = std::make_shared<MockDomainNetMap>();
+    }
+    for (int idx = 0; idx < num_gpu; ++idx) {
```
if (do_gpu)?

avilcheslopez commented on test/FFNetAgentTest.cpp:
```diff
@@ -0,0 +1,519 @@
+using testing::Throw;
+using json11::Json;
+
+class FFNetAgentTest: public :: testing :: Test
```
Potential missing tests:
* `plugin_name()`
* `policy_names()`

alawibaba committed:
- Removed NiceMock from DomainNetMapTest.
- Added tests to FFNetAgent for policy_names and plugin_name
- Switched to using GEOPM_EXPECT_THROW in TensorTwoDTest
- Added tests for set_data and == to TensorTwoDTest
- Using GEOPM_EXPECT_THROW_MESSAGE in TensorTwoDIntegrationTest

Commit:
Updated ffnet agent docs to include information about environment
variables.

avilcheslopez commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,359 @@
+            )
+        : m_platform_io(plat_io)
+          , m_last_wait{{0, 0}}
+          //, M_WAIT_SEC(0.020) // 20ms Wait
```
Remove

avilcheslopez commented on src/DomainNetMapImp.hpp:
```diff
@@ -0,0 +1,84 @@
+            };
+
+            static const std::vector<std::string> M_EXPECTED_KEYS;
+            static constexpr int M_MAX_NNET_SIZE = 1024;
```
Specify units for the max size (e.g. bytes?). Verify code currently works as expected (allows expected max. size).

avilcheslopez committed:
- Improve error checking in FFNetAgent when getting env variables
- Expand max file size for DomainNetMap config file, fix unit test
- Fixed tabs in Makefile.am

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

Commit:
- Added extra unit tests to TensorOneDTest
- Minor clean-up in NN code
- Use geopm::get_env() helper
- Rename internal structs in FFNetAgent, DomainNetMap to follow
  convention

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/docs/source/geopm_agent_ffnet.7.rst:
```diff
@@ -0,0 +1,93 @@
+
+geopm_agent_ffnet(7) -- agent for adjusting frequencies based on application behavior
```
Add link to this page from the main page?

cmcantalupo commented on test/FFNetAgentTest.cpp:
```diff
@@ -0,0 +1,543 @@
+
+        std::vector<double> m_default_policy = {0.5};
+        const std::map<std::string, double> M_REGION_CLASS = {{"dgemm", 0.75},
+                                                              {"stream",0.25}};
```
```suggestion
                                                              {"stream", 0.25}};
```

cmcantalupo commented on test/TensorOneDMatcher.hpp:
```diff
@@ -0,0 +1,50 @@
+
+using geopm::TensorOneD;
+
+class TensorOneDMatcher {
+  public:
+    using is_gtest_matcher = void;
```
WS

avilcheslopez replied:
Pardon?

avilcheslopez replied:
Oh, did you mean indentation?

cmcantalupo commented on test/TensorTwoDMatcher.hpp:
```diff
@@ -0,0 +1,68 @@
+
+using geopm::TensorTwoD;
+
+class TensorTwoDMatcher {
+  public:
+    using is_gtest_matcher = void;
```
WS

bgeltz commented on Makefile.am:
```diff
@@ -414,11 +422,17 @@ libgeopmpolicy_la_SOURCES = src/Accumulator.cpp \
                             src/MonitorAgent.hpp \
+                            src/NNFactory.hpp \
+                            src/NNFactoryImp.hpp \
+                            src/NNFactory.cpp \
```
nit
```suggestion
                            src/NNFactory.cpp \
                            src/NNFactory.hpp \
                            src/NNFactoryImp.hpp \
```

bgeltz commented on service/docs/source/geopm_agent_ffnet.7.rst:
```diff
@@ -0,0 +1,93 @@
+    to change, including being combined with other agents. Also, this agent 
+    requires a neural net json file, the format of which is described by 
+    the schema domainnetmap_neural_net.schema.json and a region hint 
+    recommender json file described by the schema
```
nit
```suggestion
    recommender JSON file described by the schema
```

bgeltz commented on service/docs/source/geopm_agent_ffnet.7.rst:
```diff
@@ -0,0 +1,93 @@
+determines the degree to which the frequency recommender is adverse to potentially
+reducing performance by reducing frequency to save energy.
+
+The neural net for region classification must be provided in a json configuration
```
nit
```suggestion
The neural net for region classification must be provided in a JSON configuration
```

bgeltz commented on service/docs/source/geopm_agent_ffnet.7.rst:
```diff
@@ -0,0 +1,93 @@
+.. literalinclude:: ../../../json_schemas/domainnetmap_neural_net.schema.json
+    :language: json
+
+The per-region frequency recommendations must be provided in a json configuration
```
```suggestion
The per-region frequency recommendations must be provided in a JSON configuration
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+            int domain_index)
+        : DomainNetMapImp(nn_path, domain_type, domain_index, platform_io(),
+                          NNFactory::make_shared()) {
+    }
```
This is the style we use for classes with passed in paramaters and initialization lists, and empty constructor bodies:
```suggestion
    DomainNetMapImp::DomainNetMapImp(const std::string &nn_path,
                                     geopm_domain_e domain_type,
                                     int domain_index)
        : DomainNetMapImp(nn_path, domain_type, domain_index, platform_io(),
                          NNFactory::make_shared())
    {
    
    }
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+            PlatformIO &plat_io,
+            std::shared_ptr<NNFactory> nn_factory)
+        : m_platform_io(plat_io)
+          , m_nn_factory(nn_factory)
```
```suggestion
    DomainNetMapImp::DomainNetMapImp(const std::string &nn_path,
                                     geopm_domain_e domain_type,
                                     int domain_index,
                                     PlatformIO &plat_io,
                                     std::shared_ptr<NNFactory> nn_factory)
        : m_platform_io(plat_io)
        , m_nn_factory(nn_factory)
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+    {
+        std::ifstream file(nn_path);
+
+        if (!file) {
```
```suggestion
        if (!file.is_open()) {
```

bgeltz commented on src/DomainNetMapImp.hpp:
```diff
@@ -0,0 +1,84 @@
+            };
+
+            static const std::vector<std::string> M_EXPECTED_KEYS;
+            static constexpr int M_MAX_NNET_SIZE_B = 1024 * 1024;
```
Why SIZE_B?  Is there a SIZE_A?
This could be more descriptive.

avilcheslopez replied:
B = Bytes :)
I added this as a result of feedback in our review meeting. The request was to specify the units. Should we spell out BYTES or should we simply put it in a comment?

bgeltz replied:
Ah, thanks for the explanation.  I'd put it in a comment:
```suggestion
            static constexpr int M_MAX_NNET_SIZE = 1024 * 1024; // In bytes
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        
+        std::vector<std::shared_ptr<DenseLayer>> layers;
```
nit
```suggestion
        std::vector<std::shared_ptr<DenseLayer> > layers;
```

avilcheslopez replied:
Q: is this for stylistic reasons? I know in earlier versions of C++ this was required, but in newer versions it is not.

bgeltz replied:
Purely stylistic since we're on c++17.  We do this else ware in the codebase, and I'd prefer we're consistent.

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+                                ": Neural net signal inputs must be strings.",
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            m_signal_inputs.push_back({
+                    m_platform_io.push_signal(
+                            input.string_value(),
+                            domain_type,
+                            domain_index),
+                    NAN});
```
```suggestion
            m_signal_inputs.push_back({m_platform_io.push_signal(input.string_value(),
                                                                 domain_type,
                                                                 domain_index),
                                       NAN});
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+                                ": Neural net delta inputs must be tuples of strings.",
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            m_delta_inputs.push_back({
+                    m_platform_io.push_signal(
+                            input[0].string_value(),
+                            domain_type,
+                            domain_index),
+                    m_platform_io.push_signal(
+                            input[1].string_value(),
+                            domain_type,
+                            domain_index),
+                    NAN, NAN, NAN, NAN});
```
```suggestion
            m_delta_inputs.push_back({m_platform_io.push_signal(input[0].string_value(),
                                                                domain_type,
                                                                domain_index),
                                      m_platform_io.push_signal(input[1].string_value(),
                                                                domain_type,
                                                                domain_index),
                                      NAN, NAN, NAN, NAN});
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+            input.signal_den_last = input.signal_den;
+            input.signal_num = m_platform_io.sample(input.batch_idx_num);
+            input.signal_den = m_platform_io.sample(input.batch_idx_den);
+            xs.push_back(
+                (input.signal_num - input.signal_num_last) /
+                (input.signal_den - input.signal_den_last)
+            );
```
```suggestion
            xs.push_back((input.signal_num - input.signal_num_last) /
                         (input.signal_den - input.signal_den_last));
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+
+    std::vector<double> DomainNetMapImp::trace_values() const
+    {
+        std::vector<double> rval(
+                m_last_output.get_data().begin(),
+                m_last_output.get_data().end());
```
```suggestion
        std::vector<double> rval(m_last_output.get_data().begin(),
                                 m_last_output.get_data().end());
```

bgeltz commented on src/DenseLayer.cpp:
```diff
@@ -0,0 +1,64 @@
+
+    DenseLayerImp::DenseLayerImp(const TensorTwoD &weights, const TensorOneD &biases) 
+	    : m_weights(weights)
+	      , m_biases(biases)
```
```suggestion
	    : m_weights(weights)
	    , m_biases(biases)
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        std::vector<std::vector<double> > vals;
```
I see the proper style else ware, like here.

bgeltz commented on src/TensorOneD.hpp:
```diff
@@ -0,0 +1,104 @@
+#ifndef TENSORONED_HPP_INCLUDE
+#define TENSORONED_HPP_INCLUDE
+
+#include <cstdint>
```
AFAICT, no special int types are used.
stdio.h is needed for the size_t definition.
```suggestion
#include <stdio.h>
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        std::size_t vec_size = obj.array_items().size();
```
This file mixes use of `size_t` and `std::size_t`.  Please pick one and be consistent.

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+        const auto nnet_properties = nnet_json.object_items();
+        const auto signal_it = nnet_properties.find("signal_inputs");
+        const auto delta_it = nnet_properties.find("delta_inputs");
+        size_t signal_inputs_size = 0;
+        size_t delta_inputs_size = 0;
```
This file mixes use of `size_t` and `std::size_t`.  Please pick one and be consistent.

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+        }
+    }
+
+    std::shared_ptr<DenseLayer> DomainNetMapImp::json_to_DenseLayer(const json11::Json &obj) const {
```
```suggestion
    std::shared_ptr<DenseLayer> DomainNetMapImp::json_to_DenseLayer(const json11::Json &obj) const
    {
```
I marked all the spots that I saw, but please check for this in all the new code.  The existing style we have in the code base is as I suggested.  The position of this open curly is mixed in this PR.

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+        return m_nn_factory->createDenseLayer(weights, biases);
+    }
+
+    TensorOneD DomainNetMapImp::json_to_TensorOneD(const json11::Json &obj) const {
```
```suggestion
    TensorOneD DomainNetMapImp::json_to_TensorOneD(const json11::Json &obj) const
    {
```

bgeltz commented on src/DomainNetMap.cpp:
```diff
@@ -0,0 +1,344 @@
+        return m_nn_factory->createTensorOneD(vals);
+    }
+
+    TensorTwoD DomainNetMapImp::json_to_TensorTwoD(const json11::Json &obj) const {
```
```suggestion
    TensorTwoD DomainNetMapImp::json_to_TensorTwoD(const json11::Json &obj) const
    {
```

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+        : m_platform_io(plat_io)
+          , m_last_wait{{0, 0}}
+          , m_do_write_batch(false)
+          , m_perf_energy_bias(0)
```
```suggestion
    FFNetAgent::FFNetAgent(PlatformIO &plat_io,
                           const PlatformTopo &topo,
                           const std::map<std::pair<geopm_domain_e, int>, std::shared_ptr<DomainNetMap> > &net_map,
                           const std::map<geopm_domain_e, std::shared_ptr<RegionHintRecommender> > &freq_recommender
            )
        : m_platform_io(plat_io)
        , m_last_wait{{0, 0}}
        , m_do_write_batch(false)
        , m_perf_energy_bias(0)
```

avilcheslopez replied:
Note: for the net_map and freq_recommender parameters, this would put those lines beyond the ~100 column limit by quite a bit (beyond what we typically tolerate).

bgeltz commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,115 @@
+#define FFNETAGENT_HPP_INCLUDE
+
+#include <vector>
+#include <string>
```
```suggestion
#include <string>
#include <cstdint>
#include <utility>
```
For `uint64_t` use and `std::pair` use.

bgeltz commented on src/DenseLayer.hpp:
```diff
@@ -0,0 +1,59 @@
+
+            TensorOneD operator()(const TensorOneD &input) const {
+                return forward(input);
+            }
```
Please define the implementation in the CPP file like TensorOneD does for all the operators.

bgeltz commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,54 @@
+
+            TensorOneD operator()(const TensorOneD &input) const {
+                return forward(input);
+            }
```
Please define the implementation in the CPP file like TensorOneD does for all the operators.

bgeltz commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,115 @@
+                int index;
+                bool operator<(const m_domain_key_s &other) const {
+                    return type < other.type || (type == other.type && index < other.index);
+                }
```
I'm not a fan of having this operator defined in the header file, but I don't think there's another way to do this since it's scoped to this struct.

bgeltz commented on src/FFNetAgent.hpp:
```diff
@@ -0,0 +1,115 @@
+            };
+
+            FFNetAgent();
+            FFNetAgent(
+                    PlatformIO &plat_io,
+                    const PlatformTopo &topo,
+                    const std::map<std::pair<geopm_domain_e, int>, std::shared_ptr<DomainNetMap> >
+                        &net_map,
+                    const std::map<geopm_domain_e, std::shared_ptr<RegionHintRecommender> >
+                        &freq_recommender
+            );
```
```suggestion
            FFNetAgent(PlatformIO &plat_io,
                       const PlatformTopo &topo,
                       const std::map<std::pair<geopm_domain_e, int>, std::shared_ptr<DomainNetMap> > &net_map,
                       const std::map<geopm_domain_e, std::shared_ptr<RegionHintRecommender> > &freq_recommender);
```

avilcheslopez replied:
Note: column limit

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+            {
+                {GEOPM_DOMAIN_PACKAGE, "GEOPM_CPU_NN_PATH"},
+                {GEOPM_DOMAIN_GPU, "GEOPM_GPU_NN_PATH"}
+            };
```
```suggestion
    const std::map<geopm_domain_e, std::string> FFNetAgent::M_NNET_ENVNAME =
        {{GEOPM_DOMAIN_PACKAGE, "GEOPM_CPU_NN_PATH"},
         {GEOPM_DOMAIN_GPU, "GEOPM_GPU_NN_PATH"}};
```
Please make this change for all these maps defined up to the constructor.

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+
+        if (freq_recommender.empty()) {
+            for (geopm_domain_e domain_type : m_domain_types) {
+                m_freq_recommender[domain_type] = RegionHintRecommender::make_shared(
+                        get_env_value(M_FREQMAP_ENVNAME.at(domain_type)),
+                        m_platform_io.read_signal(
+                            M_MIN_FREQ_SIGNAL_NAME.at(domain_type),
+                            GEOPM_DOMAIN_BOARD,
+                            0),
+                        m_platform_io.read_signal(
+                            M_MAX_FREQ_SIGNAL_NAME.at(domain_type),
+                            GEOPM_DOMAIN_BOARD,
+                            0)
+                        );
```
This block and several below have a different indentation style that what we normally use.  Please fix.  I can help on this.

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+                m_net_map[domain_key] = DomainNetMap::make_shared(
+                                get_env_value(M_NNET_ENVNAME.at(domain_key.type)),
+                                domain_key.type,
+                                domain_key.index);
```
```suggestion
                m_net_map[domain_key] =
                    DomainNetMap::make_shared(get_env_value(M_NNET_ENVNAME.at(domain_key.type)),
                                              domain_key.type,
                                              domain_key.index);
```

avilcheslopez replied:
Note: this would make line 110 go beyond the column limit (more than we typically tolerate).

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+    // Push signals and controls for future batch read/write
+    void FFNetAgent::init(int level, const std::vector<int> &fan_in, bool is_level_root)
+    {
+        for (const m_domain_key_s domain_key : m_domains) {
+            m_freq_control[domain_key].min_idx = 
+                    m_platform_io.push_control(
+                        M_MIN_FREQ_CONTROL_NAME.at(domain_key.type),
+                        domain_key.type,
+                        domain_key.index);
+            m_freq_control[domain_key].max_idx = 
+                    m_platform_io.push_control(
+                        M_MAX_FREQ_CONTROL_NAME.at(domain_key.type),
+                        domain_key.type,
+                        domain_key.index);
+            m_freq_control[domain_key].last_value = NAN;
+        }
```
```suggestion
        for (const m_domain_key_s domain_key : m_domains) {
            m_freq_control[domain_key].min_idx = 
                m_platform_io.push_control(M_MIN_FREQ_CONTROL_NAME.at(domain_key.type),
                                           domain_key.type,
                                           domain_key.index);
            m_freq_control[domain_key].max_idx = 
                m_platform_io.push_control(M_MAX_FREQ_CONTROL_NAME.at(domain_key.type),
                                           domain_key.type,
                                           domain_key.index);
            m_freq_control[domain_key].last_value = NAN;
        }
```

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+    {
+    }
+
+    bool FFNetAgent::is_all_nan(const std::vector<double> &vec)
```
I see that this was just copy/pasted from e.g. the FrequencyMapAgent.  Ok for now, but we should refactor both agents and move this into Helper.cpp.

avilcheslopez replied:
Noted

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+        m_do_write_batch = false;
+
+        for (const m_domain_key_s domain_key : m_domains) {
+            double new_freq = m_freq_recommender[domain_key.type]->recommend_frequency(
+                    m_net_map[domain_key]->last_output(),
+                    m_perf_energy_bias
+                    );
```
```suggestion
            double new_freq = 
                m_freq_recommender[domain_key.type]->recommend_frequency(m_net_map[domain_key]->last_output(),
                                                                         m_perf_energy_bias);
```

avilcheslopez replied:
Note: column limit

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+                    );
+            if (!std::isnan(new_freq)
+                && m_freq_control[domain_key].last_value != new_freq) {
+                m_platform_io.adjust(
+                        m_freq_control[domain_key].min_idx,
+                        new_freq
+                        );
+                m_platform_io.adjust(
+                        m_freq_control[domain_key].max_idx,
+                        new_freq
+                        );
```
```suggestion
                m_platform_io.adjust(m_freq_control[domain_key].min_idx,
                                     new_freq);
                m_platform_io.adjust(m_freq_control[domain_key].max_idx,
                                     new_freq);
```

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+
+    std::vector<std::pair<std::string, std::string> > FFNetAgent::report_host(void) const
+    {
+        std::vector<std::pair<std::string, std::string> > result;
+        return result;
```
```suggestion
        return {};
```

bgeltz commented on src/FFNetAgent.cpp:
```diff
@@ -0,0 +1,365 @@
+        std::vector<std::string> tracelist;
+        for (const m_domain_key_s domain_key : m_domains) {
+            for (const std::string& trace_name : m_net_map.at(domain_key)->trace_names()) {
+                tracelist.push_back(
+                        trace_name
+                        + M_TRACE_SUFFIX.at(domain_key.type)
+                        + std::to_string(domain_key.index));
```
```suggestion
                tracelist.push_back(trace_name +
                                    M_TRACE_SUFFIX.at(domain_key.type) +
                                    std::to_string(domain_key.index));
```

bgeltz commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,54 @@
+            /// 
+            /// @param [in] Vector of DenseLayers
+            static std::unique_ptr<LocalNeuralNet> make_unique(
+                    std::vector<std::shared_ptr<DenseLayer> > layers);
```
```suggestion
            static std::unique_ptr<LocalNeuralNet> make_unique(std::vector<std::shared_ptr<DenseLayer> > layers);
```

bgeltz commented on src/NNFactory.cpp:
```diff
@@ -0,0 +1,43 @@
+        return std::make_shared<NNFactoryImp>();
+    }
+
+    std::shared_ptr<LocalNeuralNet> NNFactoryImp::createLocalNeuralNet(const std::vector<std::shared_ptr<DenseLayer>> &layers) const
```
```suggestion
    std::shared_ptr<LocalNeuralNet> NNFactoryImp::createLocalNeuralNet(const std::vector<std::shared_ptr<DenseLayer> > &layers) const
```

bgeltz commented on src/NNFactory.cpp:
```diff
@@ -0,0 +1,43 @@
+        return std::make_shared<DenseLayerImp>(weights, biases);
+    }
+
+    TensorTwoD NNFactoryImp::createTensorTwoD(const std::vector<std::vector<double>> &vals) const
```
```suggestion
    TensorTwoD NNFactoryImp::createTensorTwoD(const std::vector<std::vector<double> > &vals) const
```

bgeltz commented on src/NNFactory.hpp:
```diff
@@ -0,0 +1,59 @@
+            /// @param [in] layers A shared pointer to a vector of DenseLayers
+            ///
+            /// @return Returns a shared pointer to a LocalNeuralNet instance
+            virtual std::shared_ptr<LocalNeuralNet> createLocalNeuralNet(const std::vector<std::shared_ptr<DenseLayer>> &layers) const = 0;
```
```suggestion
            virtual std::shared_ptr<LocalNeuralNet> createLocalNeuralNet(const std::vector<std::shared_ptr<DenseLayer> > &layers) const = 0;
```

bgeltz commented on src/NNFactory.hpp:
```diff
@@ -0,0 +1,59 @@
+            /// @param [in] vals Matrix doubles to fill TensorTwoD object
+            ///
+            /// @return Returns a TensorTwoD instance
+            virtual TensorTwoD createTensorTwoD(const std::vector<std::vector<double>> &vals) const = 0;
```
```suggestion
            virtual TensorTwoD createTensorTwoD(const std::vector<std::vector<double> > &vals) const = 0;
```

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+            const std::string &fmap_path,
+            int min_freq,
+            int max_freq)
+    {
```
```suggestion
    std::unique_ptr<RegionHintRecommender> RegionHintRecommender::make_unique(const std::string &fmap_path,
                                                                              int min_freq,
                                                                              int max_freq)
    {
```

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+    std::shared_ptr<RegionHintRecommender> RegionHintRecommender::make_shared(
+            const std::string &fmap_path,
+            int min_freq,
+            int max_freq)
```
```suggestion
    std::shared_ptr<RegionHintRecommender> RegionHintRecommender::make_shared(const std::string &fmap_path,
                                                                              int min_freq,
                                                                              int max_freq)
```

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+    RegionHintRecommenderImp::RegionHintRecommenderImp(const std::string &fmap_path, int min_freq,
+                                                       int max_freq)
+        : m_min_freq(min_freq)
+          , m_max_freq(max_freq)
```
```suggestion
        : m_min_freq(min_freq)
        , m_max_freq(max_freq)
```

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+        std::string buf, err;
+
+        std::ifstream ffile(fmap_path);
+        if (!ffile) {
```
```suggestion
        if (!ffile.is_open()) {
```

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+
+    double RegionHintRecommenderImp::recommend_frequency(
+            const std::map<std::string, double> &nn_output,
+            double phi) const {
```
```suggestion
    double RegionHintRecommenderImp::recommend_frequency(const std::map<std::string, double> &nn_output,
                                                         double phi) const
    {
```

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+        double zz = 0;
+        double freq = 0;
+        for (const auto &[region_name, probability] : nn_output) {
+            size_t phi_idx = static_cast<size_t>(
+                                std::floor(phi * (m_freq_map.at(region_name).size() - 1)));
```
```suggestion
            size_t phi_idx = 
                static_cast<size_t>(std::floor(phi *
                                               (m_freq_map.at(region_name).size() - 1)));
```

bgeltz commented on src/RegionHintRecommender.hpp:
```diff
@@ -0,0 +1,55 @@
+            ///
+            /// @param [in] nn_output Output from the DomainNetMap neural net evaluation, as a
+            ///             map from trace names to doubles.
+            /// @param [in] phi User-input perf-energy bias
```
Need a docstring here explaining the output.  It's in Hz, which might be unexpected.

asmaalrawi replied:
I don't think the output here is in Hz? It's a distribution of probabilities. Should be unitless. I was trying to keep this more generic and have specific implementation details in RegionHintRecommendorImp.hpp

bgeltz commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+            freq = m_min_freq;
+        }
+
+        return freq * 1e8;
```
I expected to see 1e9 here.  What units is freq in before the conversion?  ~It must be MHz for this to make sense.~
The intention is to have the output in Hz?  Is this correct?
It would be nice to see whatever we multiply by here either explained by a nearby comment or captured as a constexpr or similar instead of a magic number that only shows up when we go to return.

asmaalrawi replied:
We were using pstates in the integration scripts. We should move to Hz to maintain consistency with the rest of the codebase. I'll update integration scripts accordingly.

asmaalrawi replied:
Confirmed that there is no numerical reason to divert from SI units here. We're going forward with using Hz everywhere. Code has been changed to reflect that.

bgeltz commented on src/RegionHintRecommenderImp.hpp:
```diff
@@ -0,0 +1,36 @@
+#ifndef REGIONHINTRECOMMENDERIMP_HPP_INCLUDE
+#define REGIONHINTRECOMMENDERIMP_HPP_INCLUDE
+
+#include "geopm/json11.hpp"
```
Move to the CPP.  It's not needed in the header.

bgeltz commented on src/TensorMath.cpp:
```diff
@@ -0,0 +1,109 @@
+    {
+        TensorOneD rval(tensor.get_dim());
+        for (std::size_t idx = 0; idx < tensor.get_dim(); ++idx) {
+            rval[idx] = 1/(1 + expf(-tensor[idx]));
```
There's no error checking here on the expf() call.
https://en.cppreference.com/w/c/numeric/math/exp

Are range errors (overflow/underflow) possible?  Is a divide by 0 possible here?

If none of this is possible here, a comment should explain why this is the case.

asmaalrawi replied:
It is impossible to divide by 0 because the output of an exponential is asymptotic with 0, thus the denominator is at least 1. I think overflow is possible, though!

asmaalrawi replied:
Must clear errno

#include <errno.h>
After error is raised:
errno = 0;

cmcantalupo commented on src/RegionHintRecommender.cpp:
```diff
@@ -0,0 +1,128 @@
+            freq += exp(probability) * m_freq_map.at(region_name).at(phi_idx);
+            zz += exp(probability);
+        }
+        freq = freq / zz;
```
divide by zero check?

asmaalrawi replied:
Will add a debug message for this case as it is unexpected.

Commit:
- Added edge cases and refactored tests
- Added edge case handling for RegionHintRecommender for tiny probabilities
- Added edge case handling for sigmoids with extremely large/small
inputs
- Refactored TensorOneDTest
- Added test for copy constructor to TensorOneDTest
- Added tests for edge case handling listed above
- Added debug warning message for zz=0 case in RegionHintRecommender.

avilcheslopez committed:
Add missing NN test, sort them in Makefile

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>
